### PR TITLE
Dict conversion

### PR DESCRIPTION
**Pull request by sphh**
Looks like #5 was closed by me trying to get the changes into my branch …

1. https://github.com/sphh/upsilonconf/commit/b732840d8bd700318e92cb40fd6d0a0122e1cc55 has the ‘classic’ two-step approach with your replacement strategy:
    - `_modify_string()` (probably not the best name, so please feel free to change it)
    - The function `__replace_in_keys` can be deleted here.

2. https://github.com/sphh/upsilonconf/commit/13371c8efb07e7b1e5e8bc6ad24cca340794e934 uses regular expressions for the replacement:
    - `_modify_string()` (probably not the best name, so please feel free to change it)
    - The function `__replace_in_keys` can be deleted here.

3. https://github.com/sphh/upsilonconf/commit/a6d4ed77900e54b793cddf9879c4d731c240dbd8 optimizes the regular expression approach by compiling the replacement pattern just once.

Looks, like approach 3. is the fastest (but it needs to import the `re` module).

### CODE REVIEW

**Comment by sphh** [comment]
BTW if you decide to go with the classical 2-step approach, the two `while` loops looking for the guard and count characters should be enclosed in:
```python3
i = 0
try:
    while chr(i) in s:
        i += 1
except ValueError:
    raise ValueError("Cannot find a free character to be used as guard.")
else:
    guard = chr(i)

i += 1
try:
    while chr(i) in s:
        i += 1
except ValueError:
    raise ValueError("Cannot find a free character to be used as counter.")
else:
    count = chr(i)


```
for the highly unlikely (but nevertheless still possible) event, that the key consists of all available characters!

**Comment by hoedt** [comment]
Okay, I woke up with the following solution for independently replacing multiple patterns:

```python
def multi_replace(s: str, replacements: Sequence[Tuple[str, str]]):
    if len(replacements) == 0:
        return s

    old, new = replacements[-1]
    return new.join([multi_replace(sub, replacements[:1]) for sub in s.split(old)])
```

I think it is quite elegant, it does away with the problem of possibly invalid characters and it does not require `re`.
I'm not sure how well it performs, but I would argue that this could be considered a secondary concern.
For our use case, it could be called as follows:

```python
sorted_mods = sorted(key_mods.items(), key=lambda kv: len(kv[0]))
key = multi_replace(key, sorted_mods)
```

What do you think?

**Comment by sphh** [comment]
That certainly looks nifty!

IMHO the last line of `multi_replace()` should read (note the `:-1`) …
```python3
    return new.join([multi_replace(sub, replacements[:-1]) for sub in s.split(old)])
```
and I believe it is enough to check:
```python3
    if not replacements:
        return s
```
or just
```python3
    try:
        old, new = replacements[-1]
    except IndexError:
        return s
```

What I do not really like, is that each recursive calls makes a copy of the `replacements` list (`replacements[:-1]`).

BTW I just checked, if `re` is not loaded anyway et voilà, here it is: `json` makes heavy use of it! So as long, as `json` is loaded, `re` is already loaded memory.

Even `yaml` uses it (it's a bit hidden in `yaml.resolver`).

Hence I do not see any point in avoiding it at all costs.

**Comment by hoedt** [comment]
> IMHO the last line of `multi_replace()` should read (note the `:-1`) …

Yes, I also noticed that typo and edited my comment accordingly.

> and I believe it is enough to check:

It should be enough, but I find explicitly checking the length clearer (and it happens to be a tiny bit faster).

> What I do not really like, is that each recursive calls makes a copy of the replacements list (replacements[:-1])

Fair enough. Actually, I did forget that slicing does not simply create a "view" (as e.g. with numpy arrays). Therefore, it would probably be a good idea to apply a minor code optimisation:

```python
def multi_replace(s: str, replacements: Sequence[Tuple[str, str]]):
    if len(replacements) == 0:
        return s

    old, new = replacements[-1]
    rest = replacements[:-1]
    return new.join([multi_replace(sub, rest) for sub in s.split(old)])
```

If you really insist on the copying, it would probably also be possible to additionally pass the index of the current element as follows (untested):

```python
def multi_replace(s: str, replacements: Sequence[Tuple[str, str]], index: int = 0):
    if index == len(replacements):
        return s

    old, new = replacements[index]
    return new.join([multi_replace(sub, replacements, index + 1) for sub in s.split(old)])
```

> BTW I just checked, if `re` is not loaded anyway et voilà, here it is: `json` makes heavy use of it! So as long, as `json` is loaded, re is already loaded memory.

Does this mean that #4 is a non-issue now? As I mentioned there, it should be easy enough to load `json` lazily. Loading `argparse` (which also loads `re`) is trickier, but I can look into it as soon as we wrap this one up.

> Hence I do not see any point in avoiding it at all costs.

It's not that I want to avoid it at all costs. I just think this is a problem that (intuitively) shouldn't require the "complexity" of `re`. To be honest, I would (humbly) argue that my newest solution is actually more readable than the `re` solutions you proposed. However, this might just be me being too enthusiastic about my ideas (again).

**Comment by sphh** [comment]
> > and I believe it is enough to check:
> 
> It should be enough, but I find explicitly checking the length clearer

That all depends, where you come from. I believe, using `try: … except: …` is the ‘proper’ Pythonic way, but I have seen these explicit checks used by a lot of people who come from C, C++, Fortran etc.

> (and it happens to be a tiny bit faster).

Yes and no. `try: ` is faster than `if …`, if no exception is raised. With an exception, `try:` becomes slower. So it all depends, how often the `try:` clause is expected to fail! (And before you ask: I don't know, how often `if …` has to be called to break even with one failing `try …`.)

But we are splitting hairs here.

> > What I do not really like, is that each recursive calls makes a copy of the replacements list (replacements[:-1])
> 
> Fair enough. Actually, I did forget that slicing does not simply create a "view" (as e.g. with numpy arrays). Therefore, it would probably be a good idea to apply a minor code optimisation:
> 
> ```python
> def multi_replace(s: str, replacements: Sequence[Tuple[str, str]]):
>     if len(replacements) == 0:
>         return s
> 
>     old, new = replacements[-1]
>     rest = replacements[:-1]
>     return new.join([multi_replace(sub, rest) for sub in s.split(old)])
> ```

But now the slicing and coping is done explicitly in `rest = replacements[:-1]` (and it even needs to allocate the variable `rest`). If that is really a problem, make a copy _before_ calling `multi_replace()` the first time and then use:
```python3
    old, new = replacements.pop()
    return new.join([multi_replace(sub, replacements) for sub in s.split(old)])
```
which modifies `replacements` in place.

> Does this mean that #4 is a non-issue now?

Not really, because `json` is loaded and only needed, if you want to deal with `json` files! Otherwise it does not get used at all.

> As I mentioned there, it should be easy enough to load `json` lazily. Loading `argparse` (which also loads `re`) is trickier, but I can look into it as soon as we wrap this one up.

Couldn't `argparse` and all file loaders/savers put into sub-packages, which have to be imported explicitly by the user?

> > Hence I do not see any point in avoiding it at all costs.
> 
> It's not that I want to avoid it at all costs. I just think this is a problem that (intuitively) shouldn't require the "complexity" of `re`. To be honest, I would (humbly) argue that my newest solution is actually more readable than the `re` solutions you proposed. However, this might just be me being too enthusiastic about my ideas (again).

Well, personally I believe `re` is straight-forward: Just apply the regular expression and replace the matches with the appropriate values from `key_mods`. Nothing fancy here. And there is only one recursion involved: That one to dive into the `mapping`. With the split-join approach, you need two recursions. But then this might just be my preference, because I am not afraid of regular expression, because I have already written quite a few.

We can now go round and round in circles discussing it. In the end it is the result, which counts! So please take your pick, which you are happier with. (And maybe keep the other commented out, so that it does not get forgotten in the case, it should be switched.)

File: upsilonconf/config.py [hunk 1]
```diff
@@ -447,73 +449,83 @@ def to_dict(self, key_mods: Mapping[str, str] = None) -> Dict[str, Any]:
...
-            # Call this method recursively
-            value = __replace_in_keys(value, s, r)
-        except AttributeError:
-            # `value` is not of the mapping type
-            pass
-        dictionary[key.replace(s, r)] = value
+    # Build and compile replacement pattern
+    pattern = re.compile(
+        "|".join(
+            [re.escape(k) for k in sorted(key_mods, key=lambda k: len(k), reverse=True)]
```
Comment (hoedt) [inline upsilonconf/config.py:482]: Maybe you don't like to add variables, but I would insert a `sorted_mods` variable here for readability.

File: upsilonconf/config.py [hunk 2]
```diff
@@ -447,73 +449,83 @@ def to_dict(self, key_mods: Mapping[str, str] = None) -> Dict[str, Any]:
...
+            # Call this method recursively
+            value = __modify_keys(value, key_mods, pattern)
+        except AttributeError:
+            # `value` is not of the mapping type
+            pass
+
+        # Replace only, if there are replacements requested, otherwise just
+        # save the value
+        if key_mods:
+            dictionary[pattern.sub(lambda m: key_mods[m.group(0)], key)] = value
```
Comment (hoedt) [inline upsilonconf/config.py:527]: Is there a reason you did not use a `try`-`catch` here?

**Review by hoedt — CHANGES_REQUESTED** [review]
Could you also insert a comment that refers to this pull request (preferably including a link)?

**Comment by hoedt** [comment]
> which modifies `replacements` in place

That was how I implemented it at first, but if a pattern appears in multiple sub-strings, only the first sub-string would be replaced.

> We can now go round and round in circles discussing it.

You are right. I will merge your `re` solution. If we decide we don't want to have the module loaded, we can revert to my solution.

File: upsilonconf/config.py (hunk 2 above)
Comment (sphh) [reply upsilonconf/config.py:527]: In retrospect, yes: :upside_down_face:: I believe that more often than not, users will not bother about replacing characters. In these cases, the `if …` clause is faster than a failed `try …`. If there were a default `key_mods` argument, I would change it to a `try …` clause.

Do you want a `try …` here?

File: upsilonconf/config.py (hunk 1 above)
Comment (sphh) [reply upsilonconf/config.py]: > Maybe you don't like to add variables

Not necessarily …

Done. https://github.com/hoedt/upsilonconf/pull/6/commits/ca684a4c47386b7143060c35f41e38cfe3a85ffd

**Comment by sphh** [comment]
> Could you also insert a comment that refers to this pull request (preferably including a link)?

Sorry, I do not grasp, what you want me to do. Could you please explain? Thanks.

**Comment by hoedt** [comment]
Instead of keeping all possible implementations in the code, I think it would be cleaner to simply include a comment like

```python
# see https://github.com/hoedt/upsilonconf/pull/6 for a discussion on alternative implementations
```

File: upsilonconf/config.py (hunk 2 above)
Comment (hoedt) [reply upsilonconf/config.py:527]: Not necessarily, I was just wondering given your comment earlier :upside_down_face:

Comment (sphh) [reply upsilonconf/config.py:527]: I asked for it, didn't I? :stuck_out_tongue_winking_eye:

**Comment by sphh** [comment]
I see. It's done (https://github.com/hoedt/upsilonconf/pull/6/commits/685b4fbbd7f44c963b8383723f09f32e38e5e661).

**Comment by hoedt** [comment]
I wouldn't include it in the docstring. Just a one-line comment for developers who stumble upon it.

**Comment by sphh** [comment]
Voilà! (https://github.com/hoedt/upsilonconf/pull/6/commits/4d9833cdad58de2e520d4974c1950cb3e6759ebc)

**Review by hoedt — APPROVED** [review]
Great, thanks!

PS: there is no need to write comment for every commit you make. I also get notifications for each commit you push (and your commit message are typically informative enough) :wink:

**Comment by sphh** [comment]
> PS: there is no need to write comment for every commit you make.

Sometimes the thread becomes frayed and in the end I become confused and do not know, which commit belongs to which sub-discussion, so I started to add comments to the sub-threads including links to the commit, because they only appear at the very end. It might be over the top and maybe that's also a bad habit …

Thanks for your patience!

**Pull request merged**